torchtune/models/flamingo/_transform.py

```diff
@@ -0,0 +1,209 @@
+
+    @property
+    def vocab_size(self) -> int:
+        return self.tokenizer.vocab_size
+
+    def encode(
+        self,
+        text: str,
+        add_bos: bool = True,
+        add_eos: bool = True,
+    ) -> List[int]:
+        return self.tokenizer.encode(text=text, add_bos=add_bos, add_eos=add_eos)
+
+    def decode(
+        self,
+        token_ids: List[int],
+        truncate_at_eos: bool = True,
+        skip_special_tokens: bool = True,
+    ) -> str:
+        """
+        Decode a list of token ids into a string.
+
+        Args:
+            token_ids (List[int]): The list of token ids.
+            truncate_at_eos (bool): Whether to truncate the string at the end of
+                sequence token. Default is True.
+            skip_special_tokens (bool): Whether to show or skip special tokens in the decoded string.
+                Default is True.
+
+        Returns:
+            str: The decoded string.
+        """
+        return self.tokenizer.decode(
+            token_ids,
+            truncate_at_eos=truncate_at_eos,
+            skip_special_tokens=skip_special_tokens,
+        )
+
+    def tokenize_message(
+        self,
+        message: Message,
+        tokenize_header: bool = True,
+        tokenize_end: bool = True,
+    ) -> List[int]:
+        """
+        Tokenize a message into a list of token ids.
+
+        Args:
+            message (Message): The message to tokenize.
+            tokenize_header (bool): Whether to prepend a tokenized header to the message.
+            tokenize_end (bool): Whether to append eot or eom id at the end of the message.
+
+        Returns:
+            List[int]: The list of token ids.
+        """
+        return self.tokenizer.tokenize_message(
+            message=message,
+            tokenize_header=tokenize_header,
+            tokenize_end=tokenize_end,
+        )
+
+    def tokenize_messages(
+        self,
+        messages: List[Message],
+        add_eos: bool = True,
+    ) -> Tuple[List[int], List[bool]]:
+        """
+        Tokenize a list of messages into a list of token ids and masks.
+
+        Args:
+            messages (List[Message]): The list of messages to tokenize.
+            add_eos (bool): Wether to add the tokenizer's eos_id. Default True.
+
+        Returns:
+            Tuple[List[int], List[bool]]: The list of token ids and the list of masks.
+        """
+        return self.tokenizer.tokenize_messages(
+            messages=message,
+            add_eos=add_eos,
+        )
+
+    def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Apply image decoding and transformations to the "images" field in the sample
+        and tokenizization to the "messages" field in the sample. Also returns the
+        encoder mask.
+
+        Args:
+            sample (Mapping[str, Any]): A sample with a "tokens", "mask",
+                "encoder_input" and "encoder_mask" field to feed directly into the model.
```

Comment:
we don't expect encoder_input, encoder_mask to already be in sample, we should expect "images" though

Reply:
We do expect encoder_input to be in sample by this point, look at FlamingoTransform. I agree there is some questionable generalization here, but I think we should address that when it comes up. In summary, to allow unpacking of any arbitrary input for the encoder, we treat it as a dictionary.